File: lib/core/hlc.dart
/// Hybrid logical clock value: `<physicalMillis>-<counter>-<deviceId>`.
///
/// Ordering is physical, then counter, then deviceId (lexicographic). The
/// deviceId makes the order total, so two devices never produce equal clocks.
class Hlc implements Comparable<Hlc> {
  const Hlc({
    required this.physical,
    required this.counter,
    required this.deviceId,
  });

  final int physical;
  final int counter;
  final String deviceId;

  /// Sentinel used for entities migrated from schema v1 (no clock recorded).
  static const Hlc migrated = Hlc(physical: 0, counter: 0, deviceId: 'migrated');

  bool get isMigrated => physical == 0 && counter == 0 && deviceId == 'migrated';

  static Hlc parse(String value) {
    final first = value.indexOf('-');
    final second = value.indexOf('-', first + 1);
    if (first <= 0 || second <= first) {
      throw FormatException('Invalid HLC: $value');
    }
    return Hlc(
      physical: int.parse(value.substring(0, first)),
      counter: int.parse(value.substring(first + 1, second)),
      deviceId: value.substring(second + 1),
    );
  }

  static Hlc? tryParse(String? value) {
    if (value == null) return null;
    try {
      return parse(value);
    } on FormatException {
      return null;
    }
  }

  @override
  int compareTo(Hlc other) {
    if (physical != other.physical) return physical.compareTo(other.physical);
    if (counter != other.counter) return counter.compareTo(other.counter);
    return deviceId.compareTo(other.deviceId);
  }

  @override
  String toString() => '$physical-$counter-$deviceId';

  @override
  bool operator ==(Object other) =>
      other is Hlc &&
      other.physical == physical &&
      other.counter == counter &&
      other.deviceId == deviceId;

  @override
  int get hashCode => Object.hash(physical, counter, deviceId);
}

/// Issues monotonically increasing [Hlc] values for one device.
class HlcClock {
  HlcClock({required this.deviceId, int Function()? now, Hlc? last})
      : _now = now ?? (() => DateTime.now().toUtc().millisecondsSinceEpoch),
        _lastPhysical = last?.physical ?? 0,
        _lastCounter = last?.counter ?? 0;

  final String deviceId;
  final int Function() _now;
  int _lastPhysical;
  int _lastCounter;

  Hlc get last => Hlc(physical: _lastPhysical, counter: _lastCounter, deviceId: deviceId);

  Hlc next() {
    final wall = _now();
    if (wall > _lastPhysical) {
      _lastPhysical = wall;
      _lastCounter = 0;
    } else {
      _lastCounter += 1;
    }
    return last;
  }

  /// Folds a clock received from another device into this clock so that the
  /// next issued value is strictly greater than anything seen so far.
  void observe(Hlc remote) {
    if (remote.physical > _lastPhysical) {
      _lastPhysical = remote.physical;
      _lastCounter = remote.counter;
    } else if (remote.physical == _lastPhysical && remote.counter > _lastCounter) {
      _lastCounter = remote.counter;
    }
  }
}

File: tools/hub/src/hlc.ts
/// Hybrid logical clock value: `<physicalMillis>-<counter>-<deviceId>`.
///
/// Mirrors lib/core/hlc.dart exactly: ordering is physical, then counter,
/// then deviceId (lexicographic), which makes the order total.
export class Hlc {
  constructor(
    public readonly physical: number,
    public readonly counter: number,
    public readonly deviceId: string,
  ) {}

  /** Sentinel for entities migrated from schema v1 (no clock recorded). */
  static readonly migrated = new Hlc(0, 0, 'migrated');

  static parse(value: string): Hlc {
    const first = value.indexOf('-');
    const second = value.indexOf('-', first + 1);
    if (first <= 0 || second <= first) throw new Error(`Invalid HLC: ${value}`);
    const physical = parseIntStrict(value.slice(0, first));
    const counter = parseIntStrict(value.slice(first + 1, second));
    const deviceId = value.slice(second + 1);
    if (physical === null || physical < 0 || counter === null || counter < 0 || deviceId.length === 0) {
      throw new Error(`Invalid HLC: ${value}`);
    }
    return new Hlc(physical, counter, deviceId);
  }

  static tryParse(value: unknown): Hlc | null {
    if (typeof value !== 'string') return null;
    try {
      return Hlc.parse(value);
    } catch {
      return null;
    }
  }

  static compare(a: Hlc, b: Hlc): number {
    if (a.physical !== b.physical) return a.physical - b.physical;
    if (a.counter !== b.counter) return a.counter - b.counter;
    return compareStrings(a.deviceId, b.deviceId);
  }

  get isMigrated(): boolean {
    return this.physical === 0 && this.counter === 0 && this.deviceId === 'migrated';
  }

  toString(): string {
    return `${this.physical}-${this.counter}-${this.deviceId}`;
  }
}

/** `int.tryParse` semantics: whole string must be an integer literal. */
function parseIntStrict(value: string): number | null {
  if (!/^[+-]?\d+$/.test(value)) return null;
  const parsed = Number(value);
  return Number.isSafeInteger(parsed) ? parsed : null;
}

/** Dart's String.compareTo: UTF-16 code unit order. */
export function compareStrings(a: string, b: string): number {
  return a < b ? -1 : a > b ? 1 : 0;
}

/** Issues monotonically increasing [Hlc] values for one device. */
export class HlcClock {
  private lastPhysical: number;
  private lastCounter: number;

  constructor(
    public readonly deviceId: string,
    private readonly now: () => number = () => Date.now(),
    last?: Hlc,
  ) {
    this.lastPhysical = last?.physical ?? 0;
    this.lastCounter = last?.counter ?? 0;
  }

  get last(): Hlc {
    return new Hlc(this.lastPhysical, this.lastCounter, this.deviceId);
  }

  next(): Hlc {
    const wall = this.now();
    if (wall > this.lastPhysical) {
      this.lastPhysical = wall;
      this.lastCounter = 0;
    } else {
      this.lastCounter += 1;
    }
    return this.last;
  }

  /** Folds a remote clock in so the next issued value is strictly greater. */
  observe(remote: Hlc): void {
    if (remote.physical > this.lastPhysical) {
      this.lastPhysical = remote.physical;
      this.lastCounter = remote.counter;
    } else if (remote.physical === this.lastPhysical && remote.counter > this.lastCounter) {
      this.lastCounter = remote.counter;
    }
  }
}
